File: homework1/src/warmup.js
// Homework 1

function change(amount) {
  if (amount < 0) {
    throw new RangeError('amount cannot be negative');
  }
  const result = [];
  let remaining = amount;
  [25, 10, 5, 1].forEach((coin) => {
    result.push(Math.floor(remaining / coin));
    remaining %= coin;
  });
  return result;
} // Creds to Dr. Toal's in-class example

function stripQuotes(string) {
  if (string === '') {
    return string;
  }
  let result = string;
  result = result.replace(/\\|['"]+/g, ''); // Creds to Jackson for help with regular expressions
  return result;
}

function scramble(string) {
  if (string === '') {
    return string;
  }
  const indivChars = string.split('');
  // Creds to Jackson/Alejandro for the idea to use the Fisher-Yates Shuffle Algorithm.
  for (let charToSwap = indivChars.length - 1; charToSwap >= 1; charToSwap -= 1) {
    const randomIndex = Math.floor(Math.random() * (charToSwap + 1));
    const temp = indivChars[randomIndex];
    indivChars[randomIndex] = indivChars[charToSwap];
    indivChars[charToSwap] = temp;
  }
  const result = indivChars.join('');
  return result;
}


function powers(base, limit, callback) {
  let answer = 0;
  let exponent = 0;
  if (limit > 0) {
    while (answer < limit) {
      answer = base ** exponent;
      if (answer <= limit) {
        callback(answer);
        exponent += 1;
      } else {
        break;
      }
    }
  }
}

function* powersGenerator(base, limit) {
  let answer = 0;
  let exponent = 0;
  if (limit > 0) {
    while (true) {
      answer = base ** exponent;
      if (answer <= limit) {
        yield answer;
        exponent += 1;
      } else {
        break;
      }
    }
  }
}

function say(string1) {
  let result = string1;
  if (result == null) {
    return '';
  }
  return function a(string2) {
    if (string2 == null) {
      return result;
    }
    result += ` ${string2}`;
    return say(result);
  };
}

function interleave(arr, ...a) {
  const array1 = arr;
  const array2 = [...a];
  let longerArray = array1;
  let shorterArray = array2;
  if (array2.length > array1.length) {
    longerArray = array2;
    shorterArray = array1;
  }
  const result = [];
  for (let index = 0; index < shorterArray.length; index += 1) {
    result.push(array1[index]);
    if (array2[index] != null) {
      result.push(array2[index]);
    }
  }
  for (let index = shorterArray.length; index < longerArray.length; index += 1) {
    result.push(longerArray[index]);
  }
  return result;
}

const crypto = require('crypto');

function makeCryptoFunctions(key, algorithm) {
  return [
    function encrypt(str) {
      const encryption = crypto.createCipher(algorithm, key);
      let crypted = encryption.update(str, 'utf8', 'hex');
      crypted += encryption.final('hex');
      return crypted;
    },
    function decrypt(str) {
      const decryption = crypto.createDecipher(algorithm, key);
      let decrypted = decryption.update(str, 'hex', 'utf8');
      decrypted += decryption.final('utf8');
      return decrypted;
    },
  ];
} // Creds to csanz on GitHub for an implementation of Node's crypto module

function cylinder(spec) {
  let { radius = 1, height = 1 } = spec;

  const surfaceArea = () => (2 * (Math.PI * (radius ** 2))) + (2 * Math.PI * radius * height);
  const volume = () => Math.PI * (radius ** 2) * height;
  const widen = (factor) => { radius *= factor; };
  const stretch = (factor) => { height *= factor; };
  const toString = () => `Cylinder with radius ${radius} and height ${height}`;
  return Object.freeze({
    get radius() {
      return radius;
    },
    get height() {
      return height; // Creds to Tyler for help with getter functions
    },
    surfaceArea,
    volume,
    widen,
    stretch,
    toString,
  });
}

const rp = require('request-promise');

function randomName(spec) {
  const { gender, region } = spec;

  const options = {
    uri: 'http://uinames.com/api/',
    qs: {
      amount: 1,
      gender,
      region,
    },
    json: true,
  };

  return rp(options)
    .then(result => `${result.surname}, ${result.name}`);
} // Creds to Haley for helping me with randomName.

module.exports = {
  change,
  stripQuotes,
  scramble,
  powers,
  powersGenerator,
  say,
  makeCryptoFunctions,
  cylinder,
  interleave,
  randomName,
};

File: homework2/warmup.py
import random
import math
import json
import requests

def change(amount):
    if amount < 0:
        raise ValueError('amount cannot be negative')
    result = []
    remaining = amount
    coins = [25, 10, 5, 1]
    for c in coins:
        result.append(remaining // c)
        remaining %= c
    return tuple(result)

def strip_quotes(s):
    return s.translate(str.maketrans("", "", '\'' + '\"'))

def scramble(s):
    characters = list(s)
    j = len(characters) - 1
    while j > 0:
        i = random.randint(0, j)
        characters[i], characters[j] = characters[j], characters[i]
        j -= 1
    return ''.join(characters)

def say(first_string=''):
    if first_string == '':
        return first_string
    def say2(next_string=''):
        if next_string == '':
            return first_string
        return say(first_string + " " + next_string)
    return say2

def triples(limit):
    result = []
    for c in range(1, limit + 1):
        for b in range(1, c):
            for a in range(1, b):
                if c * c == a * a + b * b:
                    l = (a, b, c)
                    result.append(l)
    return sorted(result, key=lambda x: x[0])

def powers(base, limit):
    power = 1
    while power <= limit:
        yield power
        power *= base

def interleave(a, *b):
    first_length = len(a)
    second_length = len(b)
    maximum = max(first_length, second_length)
    result = []
    for i in range(0, maximum):
        if i < first_length:
            result.append(a[i])
        if i < second_length:
            result.append(b[i])
    return result

class Cylinder:
    def __init__(self, radius=1, height=1):
        self.radius = radius
        self.height = height

    @property
    def surface_area(self):
        return (2 * (math.pi * (self.radius ** 2))) + (2 * math.pi * self.radius * self.height)
    @property
    def volume(self):
        return math.pi * (self.radius * self.radius) * self.height

    def widen(self, factor):
        self.radius *= factor

    def stretch(self, factor):
        self.height *= factor

def make_crypto_functions():
    return

def random_name(gender, region):
    url = 'http://uinames.com/api/'
    kwargs = {
        'amount': 1,
        'gender': gender,
        'region': region
    }
    info = requests.get(url=url, params=kwargs)
    person = json.loads(info.content)
    if 'error' in person:
        raise ValueError('{"error":"Invalid gender"}')
    result = person.get('surname') + ', '   + person.get('name')
    return result
